### ISSUE TITLE
Support for training with class weights

### ISSUE THREAD

**Issue by ArdalanM** [beniz/deepdetect#202]
Hi,

Just like others ml libraries (scikit-learn, keras, etc..) it would be nice to have the possibility of weighting the importance of each class. So far, in DeepDetect, all classes are equally weighted in the cost function. Having a 'class_weight' parameter would allow the user to emphasize the importance of a class during the training.

Here is a example of how `class_weight` behave in multiclass classification (cover type prediction dataset):

``` python
from sklearn import datasets, linear_model, preprocessing, pipeline, metrics
SEED = 1
np.random.seed(SEED) # for reproducibility

# download the dataset
dataset = datasets.fetch_covtype(random_state=SEED, shuffle=True)

# Taking a sample of the whole training set
X = dataset.data[:100000]
Y = dataset.target[:100000]

# class weights examples. key refers to class label while value refers to class weight
# {1: 1, 2:1, 3:1, 4:1, 5:1, 6:1, 7:1}, means all classes are equally weighted (DD case)
class_weights = [
    {1: 1, 2:1, 3:1, 4:1, 5:1, 6:1, 7:1},
    {1: 0.1, 2:1, 3:1, 4:1, 5:1, 6:1, 7:1}, 
    {1: 10, 2:1, 3:1, 4:1, 5:1, 6:1, 7:1}
]

for cw in class_weights:
    print("Fitting with the following class_weights: ", cw)
    clf = pipeline.make_pipeline(
      preprocessing.StandardScaler(),
      linear_model.LogisticRegression(random_state=SEED, warm_start=True, class_weight=cw)
      )

    clf.fit(X, Y)
    y_pred = clf.predict(X)

    acc = metrics.accuracy_score(Y, y_pred)
    print("Multiclass accuracy: {}".format(acc))
    print(metrics.classification_report(Y, y_pred))
```

The output look like this:

```
Fitting with the following class_weights:  {1: 1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1}
Multiclass accuracy: 0.71414
             precision    recall  f1-score   support
          1       0.71      0.69      0.70     36331
          2       0.74      0.80      0.77     48794
          3       0.62      0.85      0.72      6243
          4       0.68      0.28      0.40       466
          5       0.25      0.01      0.02      1641
          6       0.41      0.08      0.13      2993
          7       0.71      0.52      0.60      3532

avg / total       0.70      0.71      0.70    100000

Fitting with the following class_weights:  {1: 0.1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1}
Multiclass accuracy: 0.63112
             precision    recall  f1-score   support
          1       0.84      0.23      0.37     36331
          2       0.61      0.94      0.74     48794
          3       0.62      0.85      0.72      6243
          4       0.68      0.28      0.40       466
          5       0.28      0.02      0.03      1641
          6       0.40      0.08      0.13      2993
          7       0.53      0.80      0.63      3532

avg / total       0.68      0.63      0.57    100000

Fitting with the following class_weights:  {1: 10, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1}
Multiclass accuracy: 0.6277
             precision    recall  f1-score   support
          1       0.54      0.93      0.69     36331
          2       0.81      0.47      0.60     48794
          3       0.63      0.85      0.72      6243
          4       0.68      0.28      0.40       466
          5       0.00      0.00      0.00      1641
          6       0.41      0.07      0.13      2993
          7       0.99      0.05      0.10      3532

avg / total       0.68      0.63      0.60    100000
```

We fitted 3 times with 3 different `class_weight` values (only changed the value for the first class).
The `class_weight` allow us to adapt the classification behavior toward the first class:
- First case: balanced behavior (precision ~ recall ~ f1)
- Second case: high precision, low recall
- Third case: high recall, low precision

**Comment by beniz** [comment]
Starting to gather information wrt Caffe:
- http://stackoverflow.com/questions/30486033/tackling-class-imbalance-scaling-contribution-to-loss-and-sgd

**Comment by beniz** [comment]
See PR #206.
